lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -1227,6 +1227,32 @@ LogicalResult XorOp::canonicalize(XorOp op, PatternRewriter &rewriter) {
   auto size = inputs.size();
   assert(size > 1 && "expected 2 or more operands");
 
+  // Check De Morgan's laws
+  if (op.isBinaryNot()) {
+    auto getNegations = [&](auto term) {
+      SmallVector<Value, 2> negations;
+      auto operands = term->getOperands();
+      negations.reserve(operands.size());
+      llvm::transform(operands, std::back_inserter(negations), [&](Value x) {
+        auto allOnes =
+            rewriter.create<hw::ConstantOp>(op.getLoc(), x.getType(), -1);
+        return rewriter.create<XorOp>(x.getLoc(), x, allOnes);
+      });
+      return negations;
+    };
+    auto sub = inputs.front();
```

Comment:
nit:
```cpp
if (op.isBinaryNot()) {
  ...
 auto sub = inputs.front();
}
```
can be written as:
```c++
Value sub;
if(matchPattern(op, m_Complement(m_Any(&sub)))){
}
```
It might be nicer to use `matchPattern` since this doesn't assume `inputs.front()` is negated.

Reply:
Thank you! Fixed in 
36a9135, I have to cast op to either `Operation *` or `Value` otherwise compiler don't know which implementation to use.
Do you know why `isBinaryNot` assume the second element is the constant? I thought it might be a convention to simplify code, but I find creating a Not with constant as the first element in SVOps.cpp:
```cpp
  auto full =
      rewriter.create<hw::ConstantOp>(op.getLoc(), op.cond().getType(), -1);
  Value ops[] = {full, op.cond()};
  auto cond =
      rewriter.createOrFold<comb::XorOp>(op.getLoc(), op.cond().getType(), ops);
```

Reply:
> Do you know why isBinaryNot assume the second element is the constant?

I guess it is convention but this is valid assumption.
mlir folder moves constants of commutative operands into trailing operands 

ref:
https://github.com/llvm/llvm-project/blob/f3c9a525198d5d713d44ef2eb5b64043fbe8d0d2/mlir/lib/Transforms/Utils/FoldUtils.cpp#L167-L172